content/blog/shop-and-thumbnails.md

```diff
@@ -0,0 +1,77 @@
+---
+id: shopware-and-thumbnails
+date: 2021-07-05T23:15:59
+title: Thumbnails + Shopware + PHP = 🙉
+author: Shyim
+author_title: Developer @ Shopware
+author_url: https://github.com/shyim
+author_image_url: https://avatars3.githubusercontent.com/u/6224096?s=460&u=18be3a2d46f07dd42fc2b6dee9b4b9b68bca28d2&v=4
+tags: [shopware]
+---
+
+All shop owners want a fast online shop with the best pagespeed results. The images in the shop are making a big difference in this stats.
+But can we archive this with just PHP?
+
+## Image generation with PHP
+
+In PHP we have two ways how we can generate images once with the `gd` extension or with `imagemagic`. Both of this extensions are not optimized to create very small images and they are slower. Also they lack on support on newer image formats like avif (PHP 8.1 will have avif in `gd`).
+The PHP Community has build packages like [spatie/image-optimizer](https://github.com/spatie/image-optimizer) to optimize the image after wards by executing serval optimziers in the cli. This approach requires to have functions like `exec`, `system` or `proc_open` available, which are disabled by default on some shared hosters and the optimizers installed in the host system. But there are new technologics like `ffi` or `wasm` in php.
+
+## Image generation using FFI / WASM
+
+Instead requiring optimizers installed on the system, we could bring it isolated as WASM or as shared library with FFI.
+This both approaches could be very fast, but unfornuately both technologics are not widespread in the PHP ecosystem.
+
+## Letting a external service do your image processing
+
+When we can't create the best optimized image with just plain PHP, why not letting our image processing work done on some external service. One of my favourite image processing systems is [imgproxy](https://imgproxy.net/). It can be even selfhosted, if wanted. We save the original images on the server or an external storage like s3 and just link with it to our image processing server with some parameters like width, height etc. The server's are doing the rest for us with the best possible image.
```

Comment:
```suggestion
When we can't create the best optimized image with just plain PHP, why not let our image processing work be done by some external service? One of my favourite image processing systems is [imgproxy](https://imgproxy.net/). It can be even selfhosted, if necessary. We save the original images on the server or an external storage like S3 and just link it to our image processing server with some parameters like width, height etc. The servers are doing the rest for us with the best possible image results.
```